cmd/kube-dns/SPEC.md

```diff
@@ -0,0 +1,140 @@
+are literals. The meaning of the values in angle brackets are defined
+below or in the description of the specific query response.
+
+`<zone>` = configured cluster domain, e.g. cluster.local
+
+`<ns>` = a Namespace
+
+All comparisons between query names and data in Kubernetes is **case-insensitive**.
+
+In the results column, the following definitions should be used for
+words in _italics_.
+
+- _hostname_
+  - For endpoints, in order of precedence:
+    - The value of the endpoint’s `hostname` field.
+    - A unique, system-assigned identifier for the endpoint. The exact format and source of this identifier is not prescribed by this specification. However, it must be possible to use this to identify a specific endpoint in the context of a Service. This is used for SRV queries below in the event no explicit endpoint hostname is defined by the above methods.
+
+- _ready_
+  - An endpoint is considered _ready_ if its address is in the `addresses` field of the EndpointSubset object, or the corresponding service has the `service.alpha.kubernetes.io/tolerate-unready-endpoints` annotation set to `true`.
+
+The supported queries and expected responses follow.
+
+- **Question:** `<service>.<ns>.svc.<zone>. IN A`
+  - If no Service named `<service>` exists in `<ns>`, the result will be `NXDOMAIN`.
+  - If a Service named `<service>` exists in `<ns>` with a ClusterIP, then the result will be a single A record with the specified name and the ClusterIP of that Service.
+    - Answer Format:
+      - `<service>.<ns>.svc.<zone>. <ttl> IN A <cluster-ip>`
+    - Answer Example:
+      - `kubernetes.default.svc.cluster.local. 4 IN A 10.3.0.1`
+  - If a Service named `<service>` exists in `<ns>` without a ClusterIP (i.e., a headless service)  and there are no _ready_ endpoints for that Service, then the result will be empty (i.e., an empty Answer section).
+  - If a Service named `<service>` exists in `<ns>` without a ClusterIP and there are _ready_ endpoints for the Service, the result will be an A record with the specified name and IP for each _ready_ endpoint.
+    - Answer Format:
+      - `<service>.<ns>.svc.<zone>. <ttl> IN A <endpoint-ip>`
+    - Answer Example:
+```
+      headless.default.svc.cluster.local. 4 IN A 10.3.0.1
+      headless.default.svc.cluster.local. 4 IN A 10.3.0.2
+      headless.default.svc.cluster.local. 4 IN A 10.3.0.3
+```
+- **Question:** `<hostname>.<service>.<ns>.svc.<zone>. IN A`
+  - If no Service named `<service>` exists in `<ns>`, the result will be `NXDOMAIN`.
+  - If a Service named `<service>` exists in `<ns>` and a _ready_ endpoint exists for that Service with _hostname_ of `<hostname>` the result will be a single A record with the IP address of that endpoint. If no such endpoint exists, the result will be `NXDOMAIN`.
+    - Answer Format:
+      - `<hostname>.<service>.<ns>.svc.<zone>. <ttl> IN A <endpoint-ip>`
+    - Answer Example:
+      - `my-pet.my-service.default.svc.cluster.local. 4 IN A 10.3.0.100`
```

Comment:
Given the hash hostname in the definition above, you should give a hash hostname example as well, as it will match the A record that SRV records `target` field refers to below without special casing.